File: Windows/BibleVerseReplacer.Windows/BibleStore.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Runtime.Serialization.Json;

namespace BibleVerseReplacer.Windows
{
    internal sealed class BibleStore
    {
        public static readonly BibleStore Instance = new BibleStore();

        private readonly Dictionary<VerseKey, BibleVerse> verseMap = new Dictionary<VerseKey, BibleVerse>();
        private readonly Dictionary<string, List<BibleVerse>> chapterMap = new Dictionary<string, List<BibleVerse>>();
        private readonly HashSet<string> chapterKeys = new HashSet<string>();
        private BiblePayload payload;

        private BibleStore()
        {
        }

        public string SourceSummary
        {
            get
            {
                return payload == null ? "未加载" : payload.DisplayName + " · " + payload.Id;
            }
        }

        public void Load()
        {
            string path = LocateBibleData();
            DataContractJsonSerializer serializer = new DataContractJsonSerializer(typeof(BiblePayload));
            using (FileStream stream = File.OpenRead(path))
            {
                payload = (BiblePayload)serializer.ReadObject(stream);
            }

            verseMap.Clear();
            chapterMap.Clear();
            chapterKeys.Clear();
            foreach (BibleVerse verse in payload.Verses)
            {
                for (int verseNumber = verse.Verse; verseNumber <= verse.EndVerse; verseNumber++)
                {
                    verseMap[new VerseKey(verse.Book, verse.Chapter, verseNumber)] = verse;
                }
                string key = ChapterKey(verse.Book, verse.Chapter);
                if (!chapterMap.ContainsKey(key))
                {
                    chapterMap[key] = new List<BibleVerse>();
                }
                chapterMap[key].Add(verse);
                chapterKeys.Add(ChapterKey(verse.Book, verse.Chapter));
            }

            foreach (List<BibleVerse> verses in chapterMap.Values)
            {
                verses.Sort(CompareVerses);
            }
        }

        public List<BibleVerse> VersesFor(ParsedReference parsedReference)
        {
            List<BibleVerse> result = new List<BibleVerse>();
            HashSet<string> seen = new HashSet<string>();

            foreach (PassageReference passage in parsedReference.Passages)
            {
                foreach (BibleVerse verse in VersesFor(passage))
                {
                    if (!seen.Contains(verse.CanonicalKey))
                    {
                        result.Add(verse);
                        seen.Add(verse.CanonicalKey);
                    }
                }
            }

            result.Sort(CompareVerses);
            return result;
        }

        public List<PassageVerseGroup> VerseGroupsFor(ParsedReference parsedReference)
        {
            List<PassageVerseGroup> result = new List<PassageVerseGroup>();
            HashSet<string> seen = new HashSet<string>();

            foreach (PassageReference passage in parsedReference.Passages)
            {
                List<BibleVerse> verses = new List<BibleVerse>();
                foreach (BibleVerse verse in VersesFor(passage))
                {
                    if (!seen.Contains(verse.CanonicalKey))
                    {
                        verses.Add(verse);
                        seen.Add(verse.CanonicalKey);
                    }
                }

                if (verses.Count > 0)
                {
                    result.Add(new PassageVerseGroup(passage, verses));
                }
            }

            return result;
        }

        public List<BibleVerse> VersesFor(PassageReference passage)
        {
            if (passage.IsWholeChapter)
            {
                return VersesForWholeChapter(passage.Book, passage.StartChapter);
            }

            if (!passage.StartVerse.HasValue || !passage.EndVerse.HasValue)
            {
                return VersesForWholeChapter(passage.Book, passage.StartChapter);
            }

            if (passage.StartChapter > passage.EndChapter)
            {
                throw new FormatException("范围顺序不正确");
            }

            List<BibleVerse> result = new List<BibleVerse>();
            HashSet<string> seen = new HashSet<string>();

            for (int chapter = passage.StartChapter; chapter <= passage.EndChapter; chapter++)
            {
                int start = chapter == passage.StartChapter ? passage.StartVerse.Value : 1;
                int end = chapter == passage.EndChapter ? passage.EndVerse.Value : LastVerseNumber(passage.Book, chapter);
                if (start > end)
                {
                    throw new FormatException("范围顺序不正确");
                }

                for (int verseNumber = start; verseNumber <= end; verseNumber++)
                {
                    BibleVerse verse;
                    if (!verseMap.TryGetValue(new VerseKey(passage.Book.Code, chapter, verseNumber), out verse))
                    {
                        if (!chapterKeys.Contains(ChapterKey(passage.Book.Code, chapter)))
                        {
                            throw new InvalidOperationException(passage.Book.ChineseName + " 第 " + chapter + " 章不存在");
                        }
                        throw new InvalidOperationException(passage.Book.ChineseName + " " + chapter + ":" + verseNumber + " 不存在");
                    }

                    if (!seen.Contains(verse.CanonicalKey))
                    {
                        result.Add(verse);
                        seen.Add(verse.CanonicalKey);
                    }
                }
            }

            return result;
        }

        public List<BibleVerse> VersesFor(VerseReference reference)
        {
            return VersesFor(new PassageReference(reference));
        }

        private static string LocateBibleData()
        {
            string baseDirectory = AppDomain.CurrentDomain.BaseDirectory;
            string bundled = Path.Combine(baseDirectory, "Resources", "Bible", "cmn-cu89s.json");
            if (File.Exists(bundled))
            {
                return bundled;
            }

            string local = Path.GetFullPath(Path.Combine(baseDirectory, "..", "..", "..", "..", "Resources", "Bible", "cmn-cu89s.json"));
            if (File.Exists(local))
            {
                return local;
            }

            throw new FileNotFoundException("找不到 Resources\\Bible\\cmn-cu89s.json");
        }

        private static string ChapterKey(string book, int chapter)
        {
            return book + "#" + chapter;
        }

        private List<BibleVerse> VersesForWholeChapter(BibleBook book, int chapter)
        {
            List<BibleVerse> verses;
            if (!chapterMap.TryGetValue(ChapterKey(book.Code, chapter), out verses))
            {
                throw new InvalidOperationException(book.ChineseName + " 第 " + chapter + " 章不存在");
            }
            return new List<BibleVerse>(verses);
        }

        private int LastVerseNumber(BibleBook book, int chapter)
        {
            List<BibleVerse> verses;
            if (!chapterMap.TryGetValue(ChapterKey(book.Code, chapter), out verses) || verses.Count == 0)
            {
                throw new InvalidOperationException(book.ChineseName + " 第 " + chapter + " 章不存在");
            }
            return verses[verses.Count - 1].EndVerse;
        }

        private static int CompareVerses(BibleVerse left, BibleVerse right)
        {
            int order = left.Order.CompareTo(right.Order);
            if (order != 0)
            {
                return order;
            }

            int chapter = left.Chapter.CompareTo(right.Chapter);
            return chapter != 0 ? chapter : left.Verse.CompareTo(right.Verse);
        }
    }
}

File: Sources/BibleVerseReplacer/BibleStore.swift
import Foundation

final class BibleStore {
    static let shared = BibleStore()

    private(set) var payload: BiblePayload?
    private var verseMap: [VerseKey: BibleVerse] = [:]
    private var chapterMap: [String: [BibleVerse]] = [:]
    private var chapterKeys: Set<String> = []

    var displayName: String {
        payload?.displayName ?? "经文库未加载"
    }

    var sourceSummary: String {
        guard let payload else {
            return "未加载"
        }
        return "\(payload.displayName) · \(payload.id)"
    }

    func load() throws {
        let dataURL = try locateBibleData()
        let data = try Data(contentsOf: dataURL)
        let decoded = try JSONDecoder().decode(BiblePayload.self, from: data)

        var nextVerseMap: [VerseKey: BibleVerse] = [:]
        var nextChapterMap: [String: [BibleVerse]] = [:]
        var nextChapterKeys: Set<String> = []
        for verse in decoded.verses {
            for verseNumber in verse.verse...verse.endVerse {
                nextVerseMap[VerseKey(book: verse.book, chapter: verse.chapter, verse: verseNumber)] = verse
            }
            nextChapterMap[chapterKey(book: verse.book, chapter: verse.chapter), default: []].append(verse)
            nextChapterKeys.insert(chapterKey(book: verse.book, chapter: verse.chapter))
        }

        payload = decoded
        verseMap = nextVerseMap
        chapterMap = nextChapterMap.mapValues { verses in
            verses.sorted { lhs, rhs in
                if lhs.order == rhs.order {
                    return lhs.verse < rhs.verse
                }
                return lhs.order < rhs.order
            }
        }
        chapterKeys = nextChapterKeys
    }

    func verses(for parsedReference: ParsedReference) throws -> [BibleVerse] {
        var result: [BibleVerse] = []
        var seen: Set<String> = []

        for passage in parsedReference.passages {
            for verse in try verses(for: passage) where !seen.contains(verse.canonicalKey) {
                result.append(verse)
                seen.insert(verse.canonicalKey)
            }
        }

        return result.sorted { lhs, rhs in
            if lhs.order == rhs.order {
                if lhs.chapter == rhs.chapter {
                    return lhs.verse < rhs.verse
                }
                return lhs.chapter < rhs.chapter
            }
            return lhs.order < rhs.order
        }
    }

    func verses(for passage: PassageReference) throws -> [BibleVerse] {
        if passage.isWholeChapter {
            return try versesForWholeChapter(book: passage.book, chapter: passage.startChapter)
        }

        guard let startVerse = passage.startVerse, let endVerse = passage.endVerse else {
            return try versesForWholeChapter(book: passage.book, chapter: passage.startChapter)
        }

        guard passage.startChapter <= passage.endChapter else {
            throw ReferenceParseError.invalidRange
        }

        var result: [BibleVerse] = []
        var seen: Set<String> = []

        for chapter in passage.startChapter...passage.endChapter {
            let start = chapter == passage.startChapter ? startVerse : 1
            let end = chapter == passage.endChapter ? endVerse : try lastVerseNumber(book: passage.book, chapter: chapter)
            guard start <= end else {
                throw ReferenceParseError.invalidRange
            }
            for verseNumber in start...end {
                let key = VerseKey(book: passage.book.code, chapter: chapter, verse: verseNumber)
                guard let verse = verseMap[key] else {
                    if !chapterKeys.contains(chapterKey(book: passage.book.code, chapter: chapter)) {
                        throw VerseLookupError.chapterNotFound(book: passage.book.chineseName, chapter: chapter)
                    }
                    throw VerseLookupError.verseNotFound(book: passage.book.chineseName, chapter: chapter, verse: verseNumber)
                }
                if !seen.contains(verse.canonicalKey) {
                    result.append(verse)
                    seen.insert(verse.canonicalKey)
                }
            }
        }

        return result
    }

    func verses(for reference: VerseReference) throws -> [BibleVerse] {
        try verses(for: PassageReference(reference: reference))
    }

    private func locateBibleData() throws -> URL {
        if let bundled = Bundle.main.url(forResource: "cmn-cu89s", withExtension: "json", subdirectory: "Bible") {
            return bundled
        }

        let workingDirectory = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
        let local = workingDirectory.appendingPathComponent("Resources/Bible/cmn-cu89s.json")
        if FileManager.default.fileExists(atPath: local.path) {
            return local
        }

        throw NSError(
            domain: "BibleVerseReplacer.BibleStore",
            code: 1,
            userInfo: [NSLocalizedDescriptionKey: "找不到 Resources/Bible/cmn-cu89s.json"]
        )
    }

    private func chapterKey(book: String, chapter: Int) -> String {
        "\(book)#\(chapter)"
    }

    private func versesForWholeChapter(book: BibleBook, chapter: Int) throws -> [BibleVerse] {
        guard let verses = chapterMap[chapterKey(book: book.code, chapter: chapter)] else {
            throw VerseLookupError.chapterNotFound(book: book.chineseName, chapter: chapter)
        }
        return verses
    }

    private func lastVerseNumber(book: BibleBook, chapter: Int) throws -> Int {
        guard let verses = chapterMap[chapterKey(book: book.code, chapter: chapter)], let last = verses.last else {
            throw VerseLookupError.chapterNotFound(book: book.chineseName, chapter: chapter)
        }
        return last.endVerse
    }
}
